README.md
# プログラミング用日本語等幅フォント Cica

## [ダウンロード](https://github.com/miiton/Cica/releases/latest)

|                                                                     |                                                                        |
| :----:                                                              | :----:                                                                 |
| **MacVim + NERDTree + CtrlP + LightLine**                           | **Cica v3との違い**                                                    |
| [![MacVim](screenshots/macvim.png)](screenshots/macvim.png)         | [![v3 and v4](screenshots/cicachange.gif)](screenshots/cicachange.gif) |
| **Julia Ready**                                                     | **GVim on Windows**                                                    |
| [![Julia on Jupyter](screenshots/julia.png)](screenshots/julia.png) | [![GVim on Windows](screenshots/gvim.png)](screenshots/gvim.png)       |



## 特徴

* 以下のフォントを合成後に調整した、プログラミング向けの等幅日本語フォントです
    - 英数字、記号類に [Hack](https://sourcefoundry.org/hack/) + [DejaVu Sans Mono](https://dejavu-fonts.github.io/)
    - それ以外の文字に [Rounded Mgen+](http://jikasei.me/font/rounded-mgenplus/)
    - [Nerd Fonts](https://github.com/ryanoasis/nerd-fonts) (= Powerlineパッチ、その他適用済。Pomiconsは除外※)
    - [Noto Emoji](https://www.google.com/get/noto/)
    - [Icons for Devs](https://github.com/mirmat/iconsfordevs)
    - RictyジェネレータからForkした各種調整
    - 独自の調整として `m` の真ん中の棒を短く（ `rn` と `m` を区別しやすく＆ちょっと可愛らしく）
* tmuxの画面分割に対応しています
* 非HiDPI（非Retina）のWindowsでも文字が欠けません
* 下付き文字や上付き文字など、特殊な文字の判読性を考慮（[JuliaLang](https://julialang.org/) を想定しています）
* ライセンス
    - フォントファイルは SIL Open Font License 1.1 です
    - 合成スクリプト等のソースコードは MIT です

```
o Cica
|\
* * Icons for Devs
|\
* * NERDFonts (without Pomicons※)
|\
* * Noto Emoji (無しバージョンあり)
|\
* * DejaVu Sans Mono
|\
* * Hack
 \
  * Rounded Mgen+
  |\
  | * 源の角ゴシック
  |
  * Rounded M+
  |
  * M+ OUTLINE FONTS
```

- ※Nerd Fontsに含まれるPomiconsのグリフは商用利用が出来ないライセンスなのでジェネレータからは除外しています。
- ※アイコン類はフォントをインストール後 [https:\/\/miiton.github.io\/Cica\/](https://miiton.github.io/Cica/) で確認出来ます。

## Q&A

### 矢印記号、半角なの...？  

ソースコード中で矢印記号はそう使わないと判断して半角幅にしました。
とはいえMarkdown中では使うこともあると思うのですが、その場合
あいまい文字が半角の環境で見切れてしまうので、それであれば
多少不格好でも判読できたほうが良いという判断です。

尚、絵文字の矢印を使うという手もあります（`U+F060 - U+F063 `）

### NerdTree、CtrlPでVueのアイコンが出ない

Vueのアイコンを追加しました（v4.1.0）

.vimrcで以下を追加すると表示できます。（要 [ryanoasis/vim\-devicons: Adds file type glyphs/icons to popular Vim plugins: NERDTree, vim\-airline, Powerline, Unite, vim\-startify and more](https://github.com/ryanoasis/vim-devicons)）

```vim
let g:WebDevIconsUnicodeDecorateFileNodesExtensionSymbols = {}
let g:WebDevIconsUnicodeDecorateFileNodesExtensionSymbols['vue'] = ''
" アイコン入力方法 : `[Ctrl+V]` > `[u]` > `e905`
let g:NERDTreeExtensionHighlightColor = {}
let g:NERDTreeExtensionHighlightColor['vue'] = '42B983'
```


## 既知の問題

* [Ubuntu 上の gVim で powerline 用フォントが少しずれます。 · Issue \#33 · miiton/Cica](https://github.com/miiton/Cica/issues/33)

## ビルド手順

### Dockerを使う場合

※Regular以外のttfファイルがおかしくなる現象が発生しています #39

```sh
git clone https://github.com/miiton/Cica.git
cd Cica
docker-compose build ; docker-compose run --rm cica  # ./dist/ に出力される
```

### 手動でやる場合

2018-08-27時点、Ubuntu 16.04 にて

```sh
sudo apt-get update
sudo apt-get -y install apt-file
sudo apt-file update
sudo apt-file search add-apt-repository
sudo apt-get -y install software-properties-common
sudo apt-get -y install fontforge unar
git clone git@github.com:miiton/Cica.git
cd Cica
curl -L https://github.com/source-foundry/Hack/releases/download/v3.003/Hack-v3.003-ttf.zip -o hack.zip
unar hack.zip
cp ttf/* sourceFonts/
rm hack.zip
rm -r ttf
curl -LO https://osdn.jp/downloads/users/8/8598/rounded-mgenplus-20150602.7z
unar rounded-mgenplus-20150602.7z
cp rounded-mgenplus-20150602/rounded-mgenplus-1m-regular.ttf ./sourceFonts
cp rounded-mgenplus-20150602/rounded-mgenplus-1m-bold.ttf ./sourceFonts
curl -L https://github.com/googlei18n/noto-emoji/raw/master/fonts/NotoEmoji-Regular.ttf -o sourceFonts/NotoEmoji-Regular.ttf
curl -LO http://sourceforge.net/projects/dejavu/files/dejavu/2.37/dejavu-fonts-ttf-2.37.zip
unar dejavu-fonts-ttf-2.37.zip
mv dejavu-fonts-ttf-2.37/ttf/DejaVuSansMono.ttf ./sourceFonts/
mv dejavu-fonts-ttf-2.37/ttf/DejaVuSansMono-Bold.ttf ./sourceFonts/
curl -L https://github.com/mirmat/iconsfordevs/raw/master/fonts/iconsfordevs.ttf -o sourceFonts/iconsfordevs.ttf
fontforge -lang=py -script cica.py
```

[fontforge のバージョンが古いと正常に動作しません #6](https://github.com/miiton/Cica/issues/6)

```
% fontforge --version

Copyright (c) 2000-2012 by George Williams.
 Executable based on sources from 14:57 GMT 31-Jul-2012-ML.
 Library based on sources from 14:57 GMT 31-Jul-2012.
fontforge 20120731
libfontforge 20120731-ML
```

## ライセンス

* [LICENSE.txt](LICENSE.txt)

## 謝辞

Cicaフォントの合成にあたり素晴らしいフォントを提供してくださっている製作者の方々に感謝いたします。

- Hack : [Hack \| A typeface designed for source code](https://sourcefoundry.org/hack/)
- Rounded Mgen+ : [自家製フォント工房](http://jikasei.me/)
- M+ OUTLINE FONTS : [M\+ FONTS](https://mplus-fonts.osdn.jp/)
- Rounded M+ : [自家製フォント工房](http://jikasei.me/)
- 源の角ゴシック : [adobe\-fonts/source\-han\-sans: Source Han Sans \| 思源黑体 \| 思源黑體 \| 源ノ角ゴシック \| 본고딕](https://github.com/adobe-fonts/source-han-sans)
- Noto Emoji : [googlei18n/noto\-emoji: Noto Emoji fonts](https://github.com/googlei18n/noto-emoji)

CHANGELOG.md
# 2019-06-26 v5.0.0

* ✨ 独自のアイコングリフを導入
* 👍 NerdFontのMaterial Design Iconsを導入しました #45 #46
* 👍 大文字シータΘ と小文字シータθ を判読しやすいようにしました

# 2019-05-27 v4.1.3

* 三点リーダーを半角に変更しました #41 #43

# 2019-04-02 v4.1.2

* U+32FFに `令和(㋿)` を追加しました #42

# 2018-09-17 v4.1.0, v4.1.1

* Icons for Devsを追加しました #37

# 2018-09-15 v4.0.0

* UbuntuMonoをHackに変更しました
* ライセンスの競合を無くし、 SIL Open Font License 1.1 になりました
* 以前のバージョンはライセンスに問題があったので非公開にしました

# 2018-01-04 v2.0.5

上半身Gopherを復活させました #20

# 🎉 2017-10-09 v2.0.0

NERDTreeなどでファイルタイプ別にアイコンを表示させる為のNerd Fontsを統合し、
v2としてリリースしました。

また、CicaEは廃止し、Cica1本にしました。

## NERDFontsを追加

詳細はこちら -> https://github.com/ryanoasis/nerd-fonts

これによりCicaE v1 から DevIcon のUnicodeが変更になりました。

## スタイル追加

Italicをそれぞれ追加し、以下の通りになりました。

- Cica Regular
- Cica RegularItalic
- Cica Bold
- Cica BoldItalic

## ambiguousグリフの変更（全角半角が曖昧なグリフ）

Rounded Mgen+、UbuntuMonoどちらも半角幅になっているグリフを
UbuntuMonoのグリフを採用するように変更しました。

| グリフ                  | 変更前        | 変更後     |
| ----                    | ----          | ----       |
| 0x00a4, # currency      | Rounded Mgen+ | UbuntuMono |
| 0x00a7, # section       | Rounded Mgen+ | UbuntuMono |
| 0x00a8, # dieresis      | Rounded Mgen+ | UbuntuMono |
| 0x00ad, # soft hyphen   | Rounded Mgen+ | UbuntuMono |
| 0x00b0, # degree        | Rounded Mgen+ | UbuntuMono |
| 0x00b1, # plus-minus    | Rounded Mgen+ | UbuntuMono |
| 0x00b4, # acute         | Rounded Mgen+ | UbuntuMono |
| 0x00b6, # pilcrow       | Rounded Mgen+ | UbuntuMono |
| 0x00d7, # multiply      | Rounded Mgen+ | UbuntuMono |
| 0x00f7, # divide        | Rounded Mgen+ | UbuntuMono |
| 0x2018, # left '        | Rounded Mgen+ | UbuntuMono |
| 0x2019, # right '       | Rounded Mgen+ | UbuntuMono |
| 0x201c, # left "        | Rounded Mgen+ | UbuntuMono |
| 0x201d, # right "       | Rounded Mgen+ | UbuntuMono |
| 0x2020, # dagger        | Rounded Mgen+ | UbuntuMono |
| 0x2021, # double dagger | Rounded Mgen+ | UbuntuMono |
| 0x2026, # ...           | Rounded Mgen+ | 変更なし   |
| 0x2122, # TM            | Rounded Mgen+ | UbuntuMono |
| 0x2191, # uparrow       | Rounded Mgen+ | 変更なし   |
| 0x2193, # downarrow     | Rounded Mgen+ | 変更なし   |


## ビルドスクリプトをPythonに

全面的にPythonに書き換えました。



cica_generator.sh
#!/bin/bash
rm -rf ./tmp # :p
rm -rf ./Cica
mkdir tmp
mkdir Cica
#
# Cica Generator
cica_version="1.0"
#
# Ricty Author: Yasunori Yusa <lastname at save.sys.t.u-tokyo.ac.jp>
# Cica Author : miiton <vo.gu.ba.miiton@gmail.com>
# This script is to generate ``Cica'' font from UbuntuMono and Rounded Mgen+ 1M.
# It requires 2-5 minutes to generate Cica. Owing to Ubuntu Font License
# Version 1.1 section 5, it is PROHIBITED to distribute the generated font.
# This script supports following versions of inputting fonts.
# * UbuntuMono Version
# * Rounded Mgen+ 1M
#
# Usage:
# 1. Install FontForge
#    Debian/Ubuntu: # apt-get install fontforge
#    Fedora/CentOS: # yum install fontforge
#    OpenSUSE:      # zypper install fontforge
#    Other Linux:   Get from http://fontforge.sourceforge.net/
# 2. Get UbuntuMono-R.ttf to ./sourceFonts/
# 3. Get rounded-mgenplus-1m-regular/bold.ttf to ./sourceFonts/ 
# 4. Run this script
#        % sh cica_generator.sh auto
# 5. Install Cica
#        % cp -f Cica*.ttf ~/.fonts/
#        % fc-cache -vf
#

# Set familyname
cica_familyname="Cica"

# Set ascent and descent (line width parameters)
cica_ascent=830
cica_descent=170

# Set bold width of ASCII glyphs
ascii_regular_width=0
ascii_bold_width=30

# Set path to fontforge command
fontforge_command="fontforge"

# Set redirection of stderr
redirection_stderr="/dev/null"

# Set fonts directories used in auto flag
fonts_directories="./sourceFonts"

# Set zenkaku space glyph
zenkaku_space_glyph=""

# Set flags
leaving_tmp_flag="false"
fullwidth_ambiguous_flag="true"
scaling_down_flag="true"

# Set filenames
modified_ubuntumono_generator="modified_ubuntumono_generator.pe"
modified_ubuntumono_regu="Modified-UbuntuMono-Regular.sfd"
modified_ubuntumono_bold="Modified-UbuntuMono-Bold.sfd"
modified_roundedmgenplus_generator="modified_roundedmgenplus_generator.pe"
modified_roundedmgenplus_regu="Modified-rounded-mgenplus-1m-regular.sfd"
modified_roundedmgenplus_bold="Modified-rounded-mgenplus-1m-bold.sfd"
cica_generator="cica_generator.pe"
post_script="post_script.pe"

########################################
# Pre-process
########################################

# Print information message
cat << _EOT_
Cica Generator ${cica_version} ( forked from Ricty Generator by Yasunori Yusa )

Author: miiton

This script is to generate \`\`Cica'' font from UbuntuMono and Rounded Mgen+ 1M.
It requires 10-60 minutes to generate Cica...

_EOT_

# Define displaying help function
cica_generator_help()
{
    echo "Usage: cica_generator.sh [options] auto"
    echo "       cica_generator.sh [options] UbuntuMono-R.ttf rounded-mgenplus-1m-regular.ttf rounded-mgenplus-1m-bold.ttf"
    echo ""
    echo "Options:"
    echo "  -h                     Display this information"
    echo "  -V                     Display version number"
    echo "  -f /path/to/fontforge  Set path to fontforge command"
    echo "  -v                     Enable verbose mode (display fontforge's warnings)"
    echo "  -l                     Leave (NOT remove) temporary files"
    echo "  -n string              Set fontfamily suffix (\`\`Cica string'')"
    echo "  -w                     Widen line space"
    echo "  -W                     Widen line space extremely"
    echo "  -b                     Make bold-face ASCII glyphs more bold"
    echo "  -B                     Make regular-/bold-face ASCII glyphs more bold"
    echo "  -Z unicode             Set visible zenkaku space copied from another glyph"
    echo "  -z                     Disable visible zenkaku space"
    echo "  -a                     Disable fullwidth ambiguous charactors"
    echo "  -s                     Disable scaling down Rounded Mgen+ 1M"
    exit 0
}

# Get options
while getopts hVf:vln:wWbBZ:zas OPT
do
    case "$OPT" in
        "h" )
            cica_generator_help
            ;;
        "V" )
            exit 0
            ;;
        "f" )
            echo "Option: Set path to fontforge command: ${OPTARG}"
            fontforge_command="$OPTARG"
            ;;
        "v" )
            echo "Option: Enable verbose mode"
            redirection_stderr="/dev/stderr"
            ;;
        "l" )
            echo "Option: Leave (NOT remove) temporary files"
            leaving_tmp_flag="true"
            ;;
        "n" )
            echo "Option: Set fontfamily suffix: ${OPTARG}"
            cica_familyname_suffix=`echo $OPTARG | tr -d ' '`
            ;;
        "w" )
            echo "Option: Widen line space"
            cica_ascent=`expr $cica_ascent + 128`
            cica_descent=`expr $cica_descent + 32`
            ;;
        "W" )
            echo "Option: Widen line space extremely"
            cica_ascent=`expr $cica_ascent + 256`
            cica_descent=`expr $cica_descent + 64`
            ;;
        "b" )
            echo "Option: Make bold-face ASCII glyphs more bold"
            ascii_bold_width=`expr $ascii_bold_width + 30`
            ;;
        "B" )
            echo "Option: Make regular-/bold-face ASCII glyphs more bold"
            ascii_regular_width=`expr $ascii_regular_width + 30`
            ascii_bold_width=`expr $ascii_bold_width + 30`
            ;;
        "Z" )
            echo "Option: Set visible zenkaku space copied from another glyph: ${OPTARG}"
            zenkaku_space_glyph="0u${OPTARG}"
            ;;
        "z" )
            echo "Option: Disable visible zenkaku space"
            zenkaku_space_glyph="0u3000"
            ;;
        "a" )
            echo "Option: Disable fullwidth ambiguous charactors"
            fullwidth_ambiguous_flag="false"
            ;;
        "s" )
            echo "Option: Disable scaling down Rounded Mgen+ 1M"
            scaling_down_flag="false"
            ;;
        * )
            exit 1
            ;;
    esac
done
shift `expr $OPTIND - 1`

# Check fontforge existance
if ! which $fontforge_command > /dev/null 2>&1
then
    echo "Error: ${fontforge_command} command not found" >&2
    exit 1
fi

# Get input fonts
if [ $# -eq 1 -a "$1" = "auto" ]
then
    # Check existance of directories
    tmp=""
    for i in $fonts_directories
    do
        [ -d "$i" ] && tmp="$tmp $i"
    done
    fonts_directories=$tmp
    # Search UbuntuMono
    input_ubuntumono=`find $fonts_directories -follow -name UbuntuMono-R.ttf | head -n 1`
    if [ -z "$input_ubuntumono" ]
    then
        echo "Error: UbuntuMono-R.ttf not found" >&2
        exit 1
    fi
    # Search Rounded Mgen+ 1M
    input_roundedmgenplus_regu=`find $fonts_directories -follow -iname rounded-mgenplus-1m-regular.ttf | head -n 1`
    input_roundedmgenplus_bold=`find $fonts_directories -follow -iname rounded-mgenplus-1m-bold.ttf    | head -n 1`
    if [ -z "$input_roundedmgenplus_regu" -o -z "$input_roundedmgenplus_bold" ]
    then
        echo "Error: rounded-mgenplus-1m-regular.ttf and/or rounded-mgenplus-1m-bold.ttf not found" >&2
        exit 1
    fi
elif [ $# -eq 3 ]
then
    # Get arguments
    input_ubuntumono=$1
    input_roundedmgenplus_regu=$2
    input_roundedmgenplus_bold=$3
    # Check existance of files
    if [ ! -r "$input_ubuntumono" ]
    then
        echo "Error: ${input_ubuntumono} not found" >&2
        exit 1
    elif [ ! -r "$input_roundedmgenplus_regu" ]
    then
        echo "Error: ${input_roundedmgenplus_regu} not found" >&2
        exit 1
    elif [ ! -r "$input_roundedmgenplus_bold" ]
    then
        echo "Error: ${input_roundedmgenplus_bold} not found" >&2
        exit 1
    fi
    # Check filename
    [ "$(basename $input_ubuntumono)" != "UbuntuMono-R.ttf" ] \
        && echo "Warning: ${input_ubuntumono} is really UbuntuMono?" >&2
    [ "$(basename $input_roundedmgenplus_regu)" != "rounded-mgenplus-1m-regular.ttf" ] \
        && echo "Warning: ${input_roundedmgenplus_regu} is really Rounded Mgen+ 1M Regular?" >&2
    [ "$(basename $input_roundedmgenplus_bold)" != "rounded-mgenplus-1m-bold.ttf" ] \
        && echo "Warning: ${input_roundedmgenplus_bold} is really Rounded Mgen+ 1M Bold?" >&2
else
    cica_generator_help
fi

# Make temporary directory
tmpdir=`mktemp -d ./tmp/cica_generator_tmpdir.XXXXXX` || exit 2


# Remove temporary directory by trapping
if [ "$leaving_tmp_flag" = "false" ]
then
    trap "if [ -d \"$tmpdir\" ]; then echo 'Remove temporary files.'; rm -rf $tmpdir; fi; echo 'Abnormally terminated.'; exit 3" HUP INT QUIT
    trap "if [ -d \"$tmpdir\" ]; then echo 'Remove temporary files.'; rm -rf $tmpdir; fi; echo 'Abnormally terminated.'" ERR
else
    trap "echo 'Abnormally terminated.'; exit 3" HUP INT QUIT
    trap "echo 'Abnormally terminated.'" ERR
fi

########################################
# Generate script for modified UbuntuMono
########################################

cat > ${tmpdir}/${modified_ubuntumono_generator} << _EOT_
#!$fontforge_command -script

# Print message
Print("Generate modified UbuntuMono.")

# Open UbuntuMono
Print("Find ${input_ubuntumono}.")
Open("${input_ubuntumono}")

# Scale to standard glyph size
ScaleToEm(860, 140)

# Remove ambiguous glyphs
if ("$fullwidth_ambiguous_flag" == "true")
    Select(0u00a4); Clear() # currency
    Select(0u00a7); Clear() # section
    Select(0u00a8); Clear() # dieresis
    Select(0u00ad); Clear() # soft hyphen
    Select(0u00b0); Clear() # degree
    Select(0u00b1); Clear() # plus-minus
    Select(0u00b4); Clear() # acute
    Select(0u00b6); Clear() # pilcrow
    Select(0u00d7); Clear() # multiply
    Select(0u00f7); Clear() # divide
    Select(0u2018); Clear() # left '
    Select(0u2019); Clear() # right '
    Select(0u201c); Clear() # left "
    Select(0u201d); Clear() # right "
    Select(0u2020); Clear() # dagger
    Select(0u2021); Clear() # double dagger
    Select(0u2026); Clear() # ...
    Select(0u2122); Clear() # TM
    Select(0u2191); Clear() # uparrow
    Select(0u2193); Clear() # downarrow
endif

# Pre-process for merging
SelectWorthOutputting()
ClearInstrs(); UnlinkReference()

# Save regular-face
Print("Save ${modified_ubuntumono_regu}.")
Save("${tmpdir}/${modified_ubuntumono_regu}")

# Make glyphs bold
Print("While making UbuntuMono bold, wait a little...")
SelectWorthOutputting()
ExpandStroke(${ascii_bold_width}, 0, 0, 0, 1)
Select(0u003e); Copy()           # >
Select(0u003c); Paste(); HFlip() # <
RoundToInt(); RemoveOverlap(); RoundToInt()

# Save bold-face
Print("Save ${modified_ubuntumono_bold}.")
Save("${tmpdir}/${modified_ubuntumono_bold}")
Close()

# Open regular-face and make it bold
if ($ascii_regular_width != 0)
    Open("${tmpdir}/${modified_ubuntumono_regu}")
    Print("While making regular-face UbuntuMono bold, wait a little...")
    SelectWorthOutputting()
    foreach
        ExpandStroke(${ascii_regular_width}, 0, 0, 0, 2)
    endloop
    Select(0u003e); Copy()           # >
    Select(0u003c); Paste(); HFlip() # <
    RoundToInt(); RemoveOverlap(); RoundToInt()
    Save("${tmpdir}/${modified_ubuntumono_regu}")
    Close()
endif

Quit()
_EOT_

########################################
# Generate script for modified Rounded Mgen+ 1M
########################################

cat > ${tmpdir}/${modified_roundedmgenplus_generator} << _EOT_
#!$fontforge_command -script

# Print message
Print("Generate modified Rounded Mgen+ 1M.")

# Set parameters
input_list  = ["${input_roundedmgenplus_regu}",    "${input_roundedmgenplus_bold}"]
output_list = ["${modified_roundedmgenplus_regu}", "${modified_roundedmgenplus_bold}"]

# Begin loop of regular and bold
i = 0
while (i < SizeOf(input_list))
    # Open Rounded Mgen+ 1M
    Print("Find " + input_list[i] + ".")
    Open(input_list[i])
    # Scale Rounded Mgen+ 1M to standard glyph size
    ScaleToEm(860, 140)
    SelectWorthOutputting()
    ClearInstrs(); UnlinkReference()
    if ("$scaling_down_flag" == "true")
        Print("While scaling " + input_list[i]:t + ", wait a little...")
        SetWidth(-1, 1); Scale(91, 91, 0, 0); SetWidth(110, 2); SetWidth(1, 1)
        Move(23, 0); SetWidth(-23, 1)
    endif
    RoundToInt(); RemoveOverlap(); RoundToInt()
    # Save modified Rounded Mgen+ 1M
    Save("${tmpdir}/" + output_list[i])
    Print("Save " + output_list[i] + ".")
    Close()
i += 1
endloop

Quit()
_EOT_

########################################
# Generate script for Cica
########################################

cat > ${tmpdir}/${cica_generator} << _EOT_
#!$fontforge_command -script

# Print message
Print("Generate Cica.")

# Set parameters
ubuntumono_list  = ["${tmpdir}/${modified_ubuntumono_regu}", \\
                     "${tmpdir}/${modified_ubuntumono_bold}"]
roundedmgenplus_list = ["${tmpdir}/${modified_roundedmgenplus_regu}", \\
                     "${tmpdir}/${modified_roundedmgenplus_bold}"]
fontfamily        = "$cica_familyname"
fontstyle_list    = ["Regular", "Bold"]
fontweight_list   = [400,       700]
panoseweight_list = [4,         7]

# Begin loop of regular and bold
i = 0
while (i < SizeOf(fontstyle_list))
    # Merge fonts
    Print("While merging " + ubuntumono_list[i]:t \\
          + " with " + roundedmgenplus_list[i]:t + ", wait a little...")
    Open(ubuntumono_list[i])
    # Copy Ubuntu Mono Glyphs (0x21-0x4f9)
    u = 0x21
    while (u < 0x4f9)
        SelectMore(u)
        u++
    endloop
    Copy()
    Close()
    Open(roundedmgenplus_list[i])
    Select(0x21)
    Paste()
    # Set configuration
    ScaleToEm(860, 140)
    SetOS2Value("Weight", fontweight_list[i]) # Book or Bold
    SetOS2Value("Width",                   5) # Medium
    SetOS2Value("FSType",                  0)
    SetOS2Value("VendorID",           "PfEd")
    SetOS2Value("IBMFamily",            2057) # SS Typewriter Gothic
    SetOS2Value("WinAscentIsOffset",       0)
    SetOS2Value("WinDescentIsOffset",      0)
    SetOS2Value("TypoAscentIsOffset",      0)
    SetOS2Value("TypoDescentIsOffset",     0)
    SetOS2Value("HHeadAscentIsOffset",     0)
    SetOS2Value("HHeadDescentIsOffset",    0)
    SetOS2Value("WinAscent",             $cica_ascent)
    SetOS2Value("WinDescent",            $cica_descent)
    SetOS2Value("TypoAscent",            693)
    SetOS2Value("TypoDescent",          -165)
    SetOS2Value("TypoLineGap",             0)
    SetOS2Value("HHeadAscent",           $cica_ascent)
    SetOS2Value("HHeadDescent",         -$cica_descent)
    SetOS2Value("HHeadLineGap",            0)
    SetPanose([2, 11, panoseweight_list[i], 9, 2, 2, 3, 2, 2, 7])
    # Edit zenkaku space (from ballot box and heavy greek cross)
    if ("$zenkaku_space_glyph" == "")
        Select(0u2610); Copy(); Select(0u3000); Paste()
        Select(0u271a); Copy(); Select(0u3000); PasteInto()
        OverlapIntersect()
    else
        Select(${zenkaku_space_glyph}); Copy(); Select(0u3000); Paste()
    endif
    # Edit zenkaku comma and period
    Select(0uff0c); Scale(150, 150, 100, 0); SetWidth(1000)
    Select(0uff0e); Scale(150, 150, 100, 0); SetWidth(1000)
    # Edit zenkaku colon and semicolon
    Select(0uff0c); Copy(); Select(0uff1b); Paste()
    Select(0uff0e); Copy(); Select(0uff1b); PasteWithOffset(0, 400)
    CenterInWidth()
    Select(0uff1a); Paste(); PasteWithOffset(0, 400)
    CenterInWidth()
    # Edit zenkaku brackets
    Select(0u0028); Copy(); Select(0uff08); Paste(); Move(250, 0); SetWidth(1000) # (
    Select(0u0029); Copy(); Select(0uff09); Paste(); Move(250, 0); SetWidth(1000) # )
    Select(0u005b); Copy(); Select(0uff3b); Paste(); Move(250, 0); SetWidth(1000) # [
    Select(0u005d); Copy(); Select(0uff3d); Paste(); Move(250, 0); SetWidth(1000) # ]
    Select(0u007b); Copy(); Select(0uff5b); Paste(); Move(250, 0); SetWidth(1000) # {
    Select(0u007d); Copy(); Select(0uff5d); Paste(); Move(250, 0); SetWidth(1000) # }
    Select(0u003c); Copy(); Select(0uff1c); Paste(); Move(250, 0); SetWidth(1000) # <
    Select(0u003e); Copy(); Select(0uff1e); Paste(); Move(250, 0); SetWidth(1000) # >
    # Edit en dash
    Select(0u2013); Copy()
    PasteWithOffset(200, 0); PasteWithOffset(-200, 0)
    OverlapIntersect()
    # Edit em dash
    Select(0u2014); Copy()
    PasteWithOffset(320, 0); PasteWithOffset(-320, 0)
    Select(0u007c); Copy(); Select(0u2014); PasteInto()
    OverlapIntersect()
    # vertical line to broken vertical line
    Select(0u00a6); Copy(); Select(0u007c); Paste() # |
    # Detach and remove .notdef
    Select(".notdef")
    DetachAndRemoveGlyphs()
    # Modify Math Symbol to 2byte
    Select(0xb1); SelectMore(0xf7)
    Scale(200,130)
    Move(250,80)
    SetWidth(1000)
    ExpandStroke(30,45,1,1,0,2)
    Select(0x2715); Copy(); Select(0xd7); Paste()
    Scale(120)
    Move(20,0)
    SetWidth(1000)
    # Post-proccess
    SelectWorthOutputting()
    if(fontstyle_list[i] == "Light")
        # Copy 0x3e -> 0x3c & HFlip
        Print("Create <")
        ExpandStroke(20,45,1,1,0,2)
        Select(0x3e); Copy(); Select(0x3c); Paste(); HFlip()
        SelectWorthOutputting()
        Print("ExpandStroke for " + fontstyle_list[i] + " style....")
    endif

    RoundToInt(); RemoveOverlap(); RoundToInt()

    SetFontNames("Cica-" + fontstyle_list[i], \\
            "Cica", \\
            "Cica " + fontstyle_list[i], \\
            fontstyle_list[i])
    SetTTFName(0x411,0, "Takahiro Minami")
    SetTTFName(0x411,1, "Cica")
    SetTTFName(0x411,2, fontstyle_list[i])
    SetTTFName(0x411,4, "Cica " + fontstyle_list[i])
    SetTTFName(0x411,16, "Cica")
    SetTTFName(0x411,17, fontstyle_list[i])
    SetTTFName(0x409,0, "Takahiro Minami")
    SetTTFName(0x409,1, "Cica")
    SetTTFName(0x409,2, fontstyle_list[i])
    SetTTFName(0x409,4, "Cica " + fontstyle_list[i])
    SetTTFName(0x409,16,"Cica")
    SetTTFName(0x409,17, fontstyle_list[i])
    # Save Cica
    Print("Save " + fontfamily + "-" + fontstyle_list[i] + ".ttf.")
    Generate("./tmp/" + fontfamily + "-" + fontstyle_list[i] + ".ttf", "", 0x84)
    Close()
i += 1
endloop

Quit()
_EOT_

########################################
# Generate Cica
########################################

$fontforge_command -script ${tmpdir}/${modified_ubuntumono_generator} \
    2> $redirection_stderr || exit 4
$fontforge_command -script ${tmpdir}/${modified_roundedmgenplus_generator} \
    2> $redirection_stderr || exit 4
$fontforge_command -script ${tmpdir}/${cica_generator} \
    2> $redirection_stderr || exit 4

# Remove temporary directory
if [ "$leaving_tmp_flag" = "false" ]
then
    echo "Remove temporary files."
    rm -rf $tmpdir
fi

echo "OS2Version to 1 Cica Regular"
. misc/os2version_reviser.sh ./tmp/Cica-Regular.ttf
echo "OS2Version to 1 Cica Bold"
. misc/os2version_reviser.sh ./tmp/Cica-Bold.ttf
# echo "OS2Version to 1 Cica Light"
# . misc/os2version_reviser.sh Cica-Light.ttf

echo "Set powerline patch to Cica Regular"
fontforge -lang=py -script ./fontpatcher/scripts/powerline-fontpatcher --no-rename ./tmp/Cica-Regular.ttf
echo "Set powerline patch to Cica Bold"
fontforge -lang=py -script ./fontpatcher/scripts/powerline-fontpatcher --no-rename ./tmp/Cica-Bold.ttf

echo "Set underline position to Cica Regular"
fontforge -lang=py -script ./modify_underline.py "./Cica Regular.ttf"
echo "Set underline position to Cica Bold"
fontforge -lang=py -script ./modify_underline.py "./Cica Bold.ttf"

mv "./Cica Regular.ttf" ./tmp/Cica-Regular.ttf
mv "./Cica Bold.ttf" ./tmp/Cica-Bold.ttf
# echo "Set powerline patch to Cica Light"
# fontforge -lang=py -script ./fontpatcher/scripts/powerline-fontpatcher --no-rename Cica-Light.ttf

########################################
# Make temporary directory
tmpdir=`mktemp -d ./tmp/post_script_tmpdir.XXXXXX` || exit 2
cat > ${tmpdir}/${post_script} << _EOT_
# Print message
Print("Post Script...")

# Modify Cica Regular
Print("Modifying Cica-Regular.ttf")
Open("./tmp/Cica-Regular.ttf")
Print("Powerline griphs")
SelectMore(0xe0a0);SelectMore(0xe0a1);SelectMore(0xe0a2)
SelectMore(0xe0b0);SelectMore(0xe0b1);SelectMore(0xe0b2);SelectMore(0xe0b3)
ClearInstrs(); UnlinkReference()
Scale(64, 100)
SetWidth(500)
Select(0xe0a0);SelectMore(0xe0a1);SelectMore(0xe0a2)
CenterInWidth()
Select(0xe0b0);SelectMore(0xe0b1)
Move(-168, -50)
SetWidth(500)
Select(0xe0b2);SelectMore(0xe0b3)
Move(-265, -50)
SetWidth(500)
RoundToInt(); RemoveOverlap(); RoundToInt()
Generate("./Cica/Cica-Regular.ttf", "", 0x84)
Close()

# Modify Cica Bold
Print("Modifying Cica-Bold.ttf")
Open("./tmp/Cica-Bold.ttf")
Print("Scaling Powerline griphs")
SelectMore(0xe0a0);SelectMore(0xe0a1);SelectMore(0xe0a2)
SelectMore(0xe0b0);SelectMore(0xe0b1);SelectMore(0xe0b2);SelectMore(0xe0b3)
ClearInstrs(); UnlinkReference()
Scale(64, 100)
SetWidth(500)
Select(0xe0a0);SelectMore(0xe0a1);SelectMore(0xe0a2)
CenterInWidth()
Select(0xe0b0);SelectMore(0xe0b1)
Move(-168, -50)
SetWidth(500)
Select(0xe0b2);SelectMore(0xe0b3)
Move(-265, -50)
SetWidth(500)
RoundToInt(); RemoveOverlap(); RoundToInt()
Generate("./Cica/Cica-Bold.ttf", "", 0x84)
Close()

_EOT_

# Execute post_script

$fontforge_command -script ${tmpdir}/${post_script} \
    2> $redirection_stderr || exit 4
# Exit
$fontforge_command -script ./02_MergeEmoji.pe
$fontforge_command -script ./03_MergeDevIcon.pe
$fontforge_command -script ./09_SetFontInfo.pe
echo "Succeeded to generate Cica! check ./Cica/ directory"
exit 0
